### PR TITLE
Feature: Background Image Containment

### PR DESCRIPTION
**Pull request by janhartmann**
This PR adds a `background-image-containment` property, which can be either set to `inside` (default) or `over`. 

When using `over` the drawing order is changed to draw background images on top of borders.

### CODE REVIEW

**Comment by janhartmann** [comment]
Closes #2726

**Comment by maxkfranz** [comment]
@janhartmann, could you reset the dist and snapshot files?  The diff is too unwieldy to review

**Comment by janhartmann** [comment]
Yes sure, no problem. I had to build it in order to for me to test it directly in my application (using the commit hash as the dependency) — but I can reset now :-)

**Comment by maxkfranz** [comment]
Great

In future, it may be easier and cleaner to use a separate branch for your builds.  The branch could be named something like ‘build’, and it’s based (or rebased) on the corresponding feature branch.  You can reference 'https://github.com/janhartmann/cytoscape.js.git#build’ in package.json to use the latest code from a ‘build’ branch.

**Comment by janhartmann** [comment]
I cleaned up the repository here - lets squash on merge.

**Comment by michalhabala** [comment]
@janhartmann  do you have, by any chance, a demo where I could see how to position the icon like that? I really like it but don't know where to start.

**Comment by janhartmann** [comment]
It can be done with a combination of this `background-image-containment` and `bounds-expansion`. Something like:

```js
{
  selector: “node”,
  style: {
    shape: "round-rectangle",
    "background-image”: [SVG_1, SVG_2],
    "background-image-containment": "over",
    "background-width": 20,
    "background-height": 20,
    "background-position-x": "122%",
    "background-position-y": -14,
    "background-clip": "none",
    "background-color": “darkblue”,
    "border-color": “blue”,
    "border-width": 1,
    "bounds-expansion": [25, 25, 0, 0],
    "text-halign": "center",
    "text-valign": "center",
    width: 65,
    height: 30
}
```

File: documentation/index.html [hunk 1]
```diff
@@ -13488,6 +13488,8 @@ <h2 id="style/background-image">Background image <a href="#style/background-imag
 </li>
 <li><p>To put an image outside of the bounds of a node’s body, it is necessary to specify <code>background-clip: none</code> and <code>bounds-expansion: n</code> for images that go <code>n</code> pixels beyond the bounding box of the node.  Note that values of <code>n</code> should be relatively small for performance.</p>
 </li>
+<li><p>To control the drawing order of background images (e.g overlay background images over borders), it is necessary to specify <code>background-image-containment: over</code> (default <code>inside</code>).</p>
```
Comment (maxkfranz) [inline documentation/index.html:13491]: Docs changes should be made in the markdown files.  The html is just generated from a template.

File: src/extensions/renderer/canvas/drawing-nodes.js [hunk 2]
```diff
@@ -264,24 +265,48 @@ CRp.drawNode = function( context, node, shiftToOriginWithBb, drawLabel = true, s
 
     context.translate( gx, gy );
 
-    setupShapeColor( ghostOpacity * bgOpacity );
+    if ( bgContainment === 'inside' ){
```
Comment (maxkfranz) [inline src/extensions/renderer/canvas/drawing-nodes.js:268]: This containment property looks great.  However, I could see some conflicts with certain use cases.  The main issue is that people would like to set this on a per-image basis, e.g. one image as the true background and one as an overlay.

This property could be per-image, like most other background image properties.  The `drawImages()` function could be parameterised to filter on only the specified subset of images.  That would keep things simple.

File: src/style/properties.js [hunk 3]
```diff
@@ -48,6 +48,7 @@ const styfn = {};
     bgFit: { enums: [ 'none', 'contain', 'cover' ], multiple: true },
     bgCrossOrigin: { enums: [ 'anonymous', 'use-credentials' ], multiple: true },
     bgClip: { enums: [ 'none', 'node' ], multiple: true },
+    bgContainment: { enums: [ 'inside', 'over' ] },
```
Comment (maxkfranz) [inline src/style/properties.js:51]: This property could be made to be `multiple` to allow for per-image values.

File: src/extensions/renderer/canvas/drawing-nodes.js (hunk 2 above)
Comment (maxkfranz) [reply src/extensions/renderer/canvas/drawing-nodes.js:268]: N.b. if only one value is specified, then that value should be applied to all images.

E.g. [inside, over, inside] v.s. inside

**Review by maxkfranz — CHANGES_REQUESTED** [review]
Looks great.  I think there are two  minor improvements that could be made:

(1) Allow the property to be specified per image.

(2) Add a homepage demo that showcases the different ways in which background images can be used.  This would show off your new feature and make it easier for library consumers to see how it can be used.  Examples could include:

- A regular background image, normally positioned (e.g. top left)
- A background image that is stretched out to cover the node
- A background image that is used like a notification icon, like your example
- An outside notification icon and a regular background image in tandem
- Other examples?

**Comment by maxkfranz** [comment]
> I cleaned up the repository here - lets squash on merge.

OK, we'll squash it

**Comment by maxkfranz** [comment]
FYI @janhartmann, I'm aiming for the 3.18 release on 2020-12-16.  Do you think (1) and (2) could be addressed by then?

**Comment by maxkfranz** [comment]
FYI @janhartmann, I'm aiming for the 3.18 release on 2020-12-16.  Do you think (1) and (2) could be addressed by then?

**Comment by maxkfranz** [comment]
@janhartmann, would you be able to address the feedback by the 21st of December so that your PR can be included in the next feature release?

**Comment by janhartmann** [comment]
Sorry for the late reply, we have a huge delivery at work that takes a lot of my time.  I should be able to pick this up soon, the changes seems fair 👍

**Comment by janhartmann** [comment]
Yes, I will need to pick this up asap. I have some time from tomorrow again.

File: src/extensions/renderer/canvas/drawing-nodes.js (hunk 2 above)
Comment (janhartmann) [reply src/extensions/renderer/canvas/drawing-nodes.js:268]: We are changing the complete drawing order based on the `background-image-containment` property:

Inside:

1. Shape Color
2. Shape
3. Image
4. Pie (deprecated)
5. Border Color
6. Border
7. Darken

Over:
1. Shape Color
2. Shape
3. Border Color
4. Border
5. Pie (deprecated)
6. Images
7. Darken

I can not see how we can introduce the multiple enum value to change the drawing order for some styles only.

**Comment by maxkfranz** [comment]
It’s not two separate drawing cases.  It’s one:

Shape Color
Shape
Images that are set to be within
Border Color
Border
Pie (deprecated)
Images that are set to be on top
Darken

**Comment by janhartmann** [comment]
You are right, had a too long vacation to think of it like this. Let me fix it up.

> Den 10. jan. 2021 kl. 20.37 skrev Max Franz <notifications@github.com>:
> 
> ﻿
> It’s not two separate drawing cases. It’s one:
> 
> Shape Color
> Shape
> Images that are set to be within
> Border Color
> Border
> Pie (deprecated)
> Images that are set to be on top
> Darken
> 
>

**Comment by janhartmann** [comment]
With the newest commit, we have the default behaviour on the right node, `inside` and the `over` on the left node

image

Code for demo:

```js
cytoscape({
  container: document.getElementById('cy'),
  boxSelectionEnabled: false,
  autounselectify: true,
  style: [{
    selector: 'node',
    style: {
      shape: 'round-rectangle',
      'background-image': ['https://live.staticflickr.com/7272/7633179468_3e19e45a0c_b.jpg', 'https://live.staticflickr.com/3063/2751740612_af11fb090b_b.jpg'],
      'background-image-containment': ['over', 'inside'],
      'background-clip': ['none', 'none'],
      'bounds-expansion': 25,
      'background-width': [20, 20],
      'background-height': [20, 20],
      'background-position-x': ['-10','120%'],
      'background-position-y': ['-10','-10'],
      'background-color': 'lightgray',
      'border-color': 'gray',
      'border-width': 1,
      'text-halign': 'center',
      'text-valign': 'center',
      width: 65,
      height: 30
    }
  }],
  elements: {
    nodes: [
      { data: { id: 'inside', name: 'Test', } },
    ]
  },
  layout: {
    name: 'grid',
    padding: 10
  }
});
```

**Comment by maxkfranz** [comment]
Looks great.  I've merged and resolved the linting issue

**Pull request merged**